### PR TITLE
[Serve] Run long poll callbacks in event loop

### LINKED ISSUE #20971: [Bug] [Serve] Ray hangs on API methods
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Serve

### What happened + What you expected to happen

After connecting to Ray and Ray Serve on a remote Ray cluster (running on k8s), running a job, and then waiting for a little while, future `serve`/`ray` methods seem to block indefinitely.

### Versions / Dependencies

ray[serve]==1.9.0
Python 3.7.12

### Reproduction script

Repro script with experiment results commented (Note: must edit remote cluster URL):

```python
import logging
import time

import ray
from ray import serve
from tqdm import tqdm

logger = logging.getLogger("ray")


def init_ray(use_remote: bool = True, verbose: bool = True):
    logger.info("Entering init_ray")
    if ray.is_initialized():
        logger.info("Ray is initialized")
        # NOTE: If you put `ray.shutdown()` here and remove the return, the script will also hang on that.
        return

    if use_remote:
        # This should be a remote ray cluster connected to with the Ray Client
        address = "ray://<your Ray client URL>:10001"
        logger.info("Running ray.init")
        ray.init(address=address, namespace="serve", log_to_driver=verbose)

        # Start Ray Serve for model serving
        # Bind on 0.0.0.0 to expose the HTTP server on external IPs.
        logger.info("Running serve.start")
        serve.start(detached=True, http_options={"host": "0.0.0.0"})


DEPLOYMENT_NAME = "DeployClass"
ray_autoscaling_config = {
    "min_replicas": 1,
    "max_replicas": 100,
    "target_num_ongoing_requests_per_replica": 5,
}


@serve.deployment(
    name=DEPLOYMENT_NAME,
    version="v1",  # required for autoscaling at the moment
    max_concurrent_queries=10,
    _autoscaling_config=ray_autoscaling_config,
)
class DeployClass:
    def f(self, i: int):
        logger.info(f"Handling {i}")
        time.sleep(2)
        return i


def deploy_deployment():
    try:
        # NOTE: This is the line it stalls on! The first `serve.` line
        logger.info("Trying to get existing deployment")
        return serve.get_deployment(DEPLOYMENT_NAME)
    except KeyError:
        logger.info("DeployClass is not currently deployed, deploying...")
        DeployClass.deploy()
        return DeployClass


inputs = list(range(10))

for i in range(5):
    logger.info("Starting ray init")
    init_ray(True, True)
    logger.info("Deploying deployment")
    deployment = deploy_deployment()
    logger.info("Getting handle")
    handle = deployment.get_handle()

    logger.info("Making method calls")
    futures = [handle.f.remote(i) for i in inputs]
    logger.info("Getting results")
    results = ray.get(futures)
    logger.info(f"Results: {results}")

    # simulate doing lots of other work...
    # Confirmed to not work:
    # 1) 10m (waited 5m on serve.get_deployment before interrupting). Also saw
    #    `Polling request timed out` error on `listen_for_changes`
    # 2) 2m (waited 10m on serve.get_deployment before interrupting). Also saw
    #    `Polling request timed out` error on `listen_for_changes`
    # 3) 1m (waited 10m on serve.get_deployment before interrupting). Also saw
    #    `Polling request timed out` error on `listen_for_changes`
    # 4) 30s (waited 10m on serve.get_deployment before interrupting). Also saw
    #    `Polling request timed out` error on `listen_for_changes`
    # Confirmed to work sometimes:
    # 5) 15s (worked 2x, then stalled out on iteration #3)
    # 5) 30s (worked 1x, then stalled out on iteration #2)
    logger.info(f"Waiting for a while...")
    for minute in tqdm(range(1)):
        logger.info(f"Waiting a minute (already waited {minute})")
        time.sleep(60)
```

### Anything else

Every time for certain wait periods. See `Confirmed to work`/`Confirmed to not work` experiments at the bottom of the repro script.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR make sure long poll client's poll_next method is always called inside an event loop as scheduled callback. This prevent the bug that occurs in Ray Client where the ObjectRef's callback cannot be used to execute other Ray command because ray client is single threaded. 

The tradeoff we are making here is that all long poll client must specify an event loop from now on. This is fine because all existing usage actually already pass event loop in. We just need to fix a test to make the enforcement.

## Related issue number
Closes #20971

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/long_poll.py [hunk 1]
```diff
@@ -104,6 +113,17 @@ def _poll_next(self):
         self._current_ref._on_completed(
```
Comment (jiaodong) [inline python/ray/serve/long_poll.py:113]: we haven't changed the fact that we're overwriting `_on_completed` when ray client is used, can this be avoided ?

**Review by jiaodong — COMMENTED** [review]
Thanks for the quick fix ! Enforcing event loop and using it to schedule and handle callbacks makes a lot of sense to me as improvement of existing implementation. I have some comments and questions about long polling in general however that maybe you had most context about:

1) Looks like we're passing around function as callbacks in multiple instances, LongestPrefixRouter.update_routes -> HTTProxy's long_poll_client instance, where both instances of PrefixRouter and LongPollClient live in the same HTTProxy instance. 

2) We're mixing event loop scheduled functions with our custom coordination and callbacks, such as `_poll_next`, `_on_callback_completed`.

As a result we have a few nested levels of call back but what we're ultimately updating are living in the same place, so do we really need that many callbacks here ? I'm curious why we haven't done it either directly update values on the LongestPrefixRouter instance, or completely use asyncio event loop + futures for coordination ?

File: python/ray/serve/long_poll.py (hunk 1 above)
Comment (simon-mo) [reply python/ray/serve/long_poll.py:113]: That's a bigger change unfortunately

**Comment by simon-mo** [comment]
This actually simplifies the callback flow. Here's what's happening with this PR:
- A long poll client is created with specified python event loop and a list of callbacks.
- The long poll client itself doesn't requires a python event loop. It just register synchronous callback to ObjectRef for Ray Corework to invoke upon ObjectRef completion.
- Whenever the ObjectRef is completed, the synchronous callback will schedule a single function to the python event loop.
- This function will execute all the callbacks and schedule the next iteration.

So there is really just two callbacks we need to care about:
- The callback passed to object_ref.on_complete. This is necessary because long poll and ServeHandle should work within synchronous driver process and shouldn't block. 
- The callback sent to event loop that runs all user's callbacks and poll next. This must run inside the python event loop because user's callback expect that.

HOWEVER, since we are now making assumption that a python event loop must exist, we can simplifies all these callback using asyncio's task and future system (similar to ServeController). But that's a much bigger change and more risky refactoring. (i still want to do that!)

**Review by jiaodong — APPROVED** [review]
synced offline and accepting to unblock current P0 issue. 

With more context from Simon, we wrote most of long polling code using callbacks while we don't have an event loop in serve, which lead to where we are right now. This PR includes minimal change of existing code to unblock user issue, while proper long term solution is to refactor long polling code to use event loop functions .. subject to discussion in later planning.

**Pull request merged**